### PR TITLE
named key - PUT and DELETE support

### PR DESCRIPTION
**Pull request by marcparadise**
Combined previous DELETE PR with PUT support. 

Pending are tests for  PUT authorization - held off because those are going to b identical to the DELETE auth tests, and I'd like to find a way to do that sanely.

### CODE REVIEW

File: lib/pedant/platform.rb [hunk 1]
```diff
@@ -345,21 +345,22 @@ def create_user(username, options = {})
...
+      if options.has_key?(:overrides)
+        payload = payload.merge(options[:overrides])
       end
+      users_url = "#{@server}/users"
+      post(users_url, @superuser, :payload => payload)
 
+    end
+    def create_user(username, options = {})
+      r = create_min_user(username, options)
+      raise "Could not create user #{username}: #{r}" unless r.code == 201
```
Comment (tylercloke) [inline lib/pedant/platform.rb]: I get worried at some point about our test helpers needing their own tests...what will be the effect of raising here?

File: spec/api/keys/keys_spec.rb [hunk 2]
```diff
@@ -584,7 +572,7 @@ def make_user_payload(payload)
         "password" => "client-password",
         "public_key" => user['public_key']
       }
-      post("#{platform.server}/users", superuser, :payload => payload)
+      post("#{platform.server}/users", superuser, payload: payload)
```
Comment (tylercloke) [inline spec/api/keys/keys_spec.rb]: Out of curiosity, what is the advantage of doing things this way?

File: spec/api/keys/keys_spec.rb [hunk 3]
```diff
@@ -667,25 +632,34 @@ def make_user_payload(payload)
...
-      post("#{platform.server}/organizations", superuser, :payload => $other_org_payload).should look_like({:status => 201})
+      # This user is used in more tests than not, so create it for the full context.
+      # Leave it up to the tests to determine if they need it associated
+      platform.create_min_user(org_user_name, overrides: org_user_payload).should look_like(status: 201 )
+      platform.create_min_user(org_admin_name, overrides: org_admin_payload).should look_like(status: 201 )
+      platform.associate_user_with_org($org_name, org_admin_user)
+      platform.add_user_to_group($org_name, org_admin_user, "admins")
+      platform.create_org(other_org_name)
+      platform.create_min_user(other_org_user_name, overrides: other_org_user_payload).should look_like(status: 201)
+      platform.associate_user_with_org(other_org_name, other_org_user)
```
Comment (tylercloke) [inline spec/api/keys/keys_spec.rb]: I think we need to do a better job of tightly coupling state to tests (aka use context blocks in a smarter way). That would have the added benefit of speeding up the tests.

File: spec/api/keys/keys_spec.rb [hunk 4]
```diff
@@ -739,98 +731,477 @@ def make_user_payload(payload)
...
+          "when public key is invalid" => {:replace => { "public_key" => "Nope."}, :code => 400},
+          "when a key of the same name already exists" => {:replace => {"name" => "default"}, :code => 409}
+        }.each do |desc, setup|
+          it "#{desc} it responds with #{setup[:code]}" do
+            setup = {:replace=>{}, :delete => []}.merge(setup)
+            payload = key_payload.merge(setup[:replace])
+            put(@named_key_url, superuser, payload:payload).should look_like(status: setup[:code])
+          end
+        end
+      end
```
Comment (tylercloke) [inline spec/api/keys/keys_spec.rb]: I'm not a huge fan of thing from a readability / spec-ish perspective, but I guess it gets the job done.

File: spec/api/keys/keys_spec.rb [hunk 5]
```diff
@@ -739,98 +731,477 @@ def make_user_payload(payload)
...
+
+        it "to a key for an invalid actor fails with 404" do
+          put("#{org_base_url}/clients/bobclient/keys/default", superuser, payload: key_payload).should look_like(status: 404)
+        end
+
+        it "to a key for a valid actor, but the key does not exist" do
+          put("#{org_base_url}/clients/#{org_client_name}/keys/badkey", superuser, payload: key_payload).should look_like(status: 404)
+        end
+
+        context "as..." do
```
Comment (tylercloke) [inline spec/api/keys/keys_spec.rb]: Can we expand this context? I have no idea what `as...` means when reading the code. The context blocks are as important for readability as they are for the output of the tests.

**Comment by tylercloke** [comment]
Looks good overall. I just have some minor stylistic / readability issues. Also, we should make sure to document in the readme any conventions we added to rspec such as `shared`.

File: spec/api/keys/keys_spec.rb (hunk 2 above)
Comment (marcparadise) [inline spec/api/keys/keys_spec.rb]: A few less characters of typing.  It also seems to be the style we're adopting elsewhere (omnibus libs, etc).

**Pull request merged**